### ISSUE TITLE
Issues with gpu using SAC+GAIL

### ISSUE THREAD

**Issue by cwj22** [HumanCompatibleAI/imitation#655]
Hi! I'm having some trouble running SAC with GAIL if I use gpu. If I change the learner to SAC in  `docs/tutorials/3_train_gail.ipynb` it throws an error.
```

Traceback (most recent call last):
  File "/home/catherine/anaconda3/envs/py38_GAIL/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/catherine/Documents/Research/imitation/src/imitation/util/networks.py", line 28, in training_mode
    yield m
  File "/home/catherine/Documents/Research/imitation/src/imitation/algorithms/adversarial/common.py", line 458, in train
    self.train_disc()
  File "/home/catherine/Documents/Research/imitation/src/imitation/algorithms/adversarial/common.py", line 351, in train_disc
    for batch in batch_iter:
  File "/home/catherine/Documents/Research/imitation/src/imitation/algorithms/adversarial/common.py", line 608, in _make_disc_train_batches
    log_policy_act_prob = self._get_log_policy_act_prob(obs_th, acts_th)
  File "/home/catherine/Documents/Research/imitation/src/imitation/algorithms/adversarial/common.py", line 512, in _get_log_policy_act_prob
    scaled_acts_th = self.policy.scale_action(acts_th)
  File "/home/catherine/anaconda3/envs/py38_GAIL/lib/python3.8/site-packages/stable_baselines3/common/policies.py", line 364, in scale_action
    return 2.0 * ((action - low) / (high - low)) - 1.0
  File "/home/catherine/anaconda3/envs/py38_GAIL/lib/python3.8/site-packages/torch/_tensor.py", line 955, in __array__
    return self.numpy()
TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
```

python 3.8 on Ubuntu 20.04.3 LTS

Thanks in advance!

**Comment by AdamGleave** [comment]
Assigning to @ernestum to investigate & triage.

We should consider adding some tests in our CI to catch things like this, right now our CI run is CPU only making it easy to slip past, we might be able to mock a GPU just to catch device type errors.

**Comment by ernestum** [comment]
In this line we call `policy.scale_action()` on a pytorch tensor:

https://github.com/HumanCompatibleAI/imitation/blob/7485ae19f7de1eabc566897f387b4475cbfa3d9e/src/imitation/algorithms/adversarial/common.py#L504

In our case that is a `stable_baselines3.common.policies.BasePolicy` and it requires a `numpy.ndarray` as the action passed to `scale_action()`:

https://github.com/DLR-RM/stable-baselines3/blob/6b8905acdb27e7b2acd036efc918818cc7ee3e20/stable_baselines3/common/policies.py#L362

As long as we are on CPU that all works out but when we are on the GPU the tensor is not automatically converted to a numpy array any more.

The obvious fix, which would involve converting the actions to a `numpy.ndarray` before passing them to `scale_action()` does not work since it causes follow-up issues. The `scaled_acts_th` is then a `numpy.ndarray` and can't be passed on to downstream functions as a tensor.

According to my limited understanding of PyTorch we have two options:
1. Fix this in the base policy by ensuring the `low` and `high` arrays are tensors on the right device.
2. Converting to numpy before passing to `scale_action` and converting back to tensor on the right device in our `AdversarialTrainer`

Is that correct @AdamGleave ?

I will now prepare a PR with a test to reproduce this bug. Then we can decide how best to fix it.

**Comment by AdamGleave** [comment]
The 2nd option seems better, the first option would require changing the types in both Gym and SB3.

**Comment by ernestum** [comment]
Ok I will implement option 2. Why do you think a change in the gym would be necessary?

**Comment by AdamGleave** [comment]
The action space is a `gym.Space` I think, and low/high are defined as NumPy arrays there IIRC.

**Comment by ernestum** [comment]
Ah I see.

**Comment by ernestum** [comment]
Fixed by #660
